spec/services/provider_details_creator_spec.rb

```diff
@@ -4,177 +4,297 @@
+        stub_provider_details_retriever(
+          provider:,
+          firm: ccms_firm,
+          offices: [ccms_office],
+        )
 
-    it "creates the right offices" do
-      expect { subject }.to change(Office, :count).by(2)
-      expect(firm.offices.count).to eq(2)
-      expect(office1.code).to eq(ccms_office1.code)
-      expect(office2.code).to eq(ccms_office2.code)
-    end
+        described_class.call(provider)
 
-    it "updates the name of the provider" do
-      expect { subject }.to change { provider.reload.name }.to("")
-    end
+        firm_roles = provider.firm.permissions.collect(&:role)
+        expect(firm_roles).to contain_exactly(
+          "application.passported.*",
+          "application.non_passported.*",
+        )
+      end
 
-    it "updates the contact_id of the provider" do
-      expect { subject }.to change { provider.reload.contact_id }.to(contact_id)
-    end
+      it "creates the offices" do
+        ccms_firm = FirmStruct.new(1, "Test Firm")
+        ccms_office1 = OfficeStruct.new(1, "6D424Y")
+        ccms_office2 = OfficeStruct.new(2, "4C880Q")
+        provider = create(:provider, username: "test-user")
+        stub_provider_details_retriever(
+          provider:,
+          firm: ccms_firm,
+          offices: [ccms_office1, ccms_office2],
+        )
+
+        described_class.call(provider)
 
-    context "when the names match" do
-      it "updates the contact_id of the provider" do
-        expect { subject }.to change { provider.reload.contact_id }.to(api_response[:contacts][0][:id])
+        offices = provider.firm.offices
+        expect(offices).to contain_exactly(
+          have_attributes(
+            class: Office,
+            ccms_id: ccms_office1.id.to_s,
+            code: ccms_office1.code,
+          ),
+          have_attributes(
+            class: Office,
+            ccms_id: ccms_office2.id.to_s,
+            code: ccms_office2.code,
+          ),
+        )
       end
-    end
 
-    context "when the username matches" do
-      subject { described_class.call(third_provider) }
+      it "updates the provider" do
+        ccms_firm = FirmStruct.new(1, "Test Firm")
+        ccms_office = OfficeStruct.new(1, "6D424Y")
+        provider = create(:provider, name: nil, username: "test-user")
+        stub_provider_details_retriever(
+          provider:,
+          firm: ccms_firm,
+          offices: [ccms_office],
+        )
 
-      before { allow(ProviderDetailsRetriever).to receive(:call).with(third_provider.username).and_return(api_response) }
+        described_class.call(provider)
 
-      it "updates the contact_id of the provider" do
-        expect { subject }.to change { third_provider.reload.contact_id }.to(api_response[:contacts][2][:id])
+        expect(provider).to have_attributes(
+          name: "",
+          contact_id: 104,
+        )
       end
     end
 
-    context "when the emails match" do
```

Comment:
This test did not test what it said it did.

Provider details are matched by `username` comparison, not email.